lib/gc/Transforms/OneDNNGraphToLinalg.cpp

```diff
@@ -46,22 +71,39 @@ Value createBroadcastOperand(Location loc, PatternRewriter &rewriter,
     return op;
   } else {
     // get broadcast dimensions
+    llvm::SmallVector<int64_t> keepDims;
     llvm::SmallVector<int64_t> bcastDims;
+    llvm::SmallVector<int64_t> collapseShape;
     for (int64_t i = 0; i < (int64_t)bcastShape.size(); i++) {
       int64_t idxOp = i - diff;
       if (idxOp < 0) {
         bcastDims.push_back(i);
       } else if (bcastShape[i] != opShape[idxOp]) {
+        assert(opShape[idxOp] == 1);
         bcastDims.push_back(i);
+      } else {
+        keepDims.push_back(idxOp);
+        collapseShape.push_back(opShape[idxOp]);
       }
     }
+    // If dimensions with size 1 need to be broadcasted, need to collapse
+    Value newVal = op;
+    if (collapseShape.size() < opShape.size()) {
+      assert(collapseShape.size() + bcastDims.size() == bcastShape.size());
+      auto reassociation =
+          computeReassociationByAnchor(keepDims, opTy.getRank());
+      ShapedType collapseTy =
+          RankedTensorType::get(collapseShape, opTy.getElementType());
+      newVal = rewriter.create<tensor::CollapseShapeOp>(loc, collapseTy, newVal,
+                                                        reassociation);
+    }
```

Comment:
consider kept dims [a, b,..., c] as anchor, so reassociation = [[...a...], [...b...], ..., [...c...]]  
e.g. for shape [16, 1, 32, 1, 64], rank = 5, kept dims = [0, 2, 4]  
[16, 1, 32, 1, 64] --collapse-> [16, 32, 64]  
reassociation = [[0], [1, 2], [3, 4]]

Reply:
added comment

Reply:
Thanks. Feels like this can still be better in the following sense. So first, we introduce an "anchor" which is the result of the collapse transformation. Kept dims is an array of indices of the shape. Finally, reassociation would be better off with a clear definition. All these are not obvious and make you read the code to understand. For example, what would happen to `[16, 1, 1, 32]`, rank `4`, kept dims `[0, 3]`? One would expect a `[16, 32]`. What should the reassociation look like?
So, for anchor - do we need another term at all? It overloads the term for fusion as well. Kept dims are clear I think. The reassociation we should also clarify.

Reply:
for [16, 1, 1, 32] -> [16, 32], rank = 4, kept dims = [0, 3]
reassociation = [[0], [1, 2, 3]]  
`anchor_dims` is just a internal term/var name for this file, to represent which dims to collapse to/expand from, I don't think it will conflict with the term for fusion.
`reassociation` is not introduced by us, it is the attribute of `tensor.collapse_shape` and `tensor.expand_shape` from [MLIR tensor dialect](https://mlir.llvm.org/docs/Dialects/TensorOps/#tensorcollapse_shape-tensorcollapseshapeop)